File: expense-app/src/app/shared/services/auth.service.ts
import { Injectable, signal, computed, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Router } from '@angular/router';
import { User, UserRole } from '../models/expense.model';
import { environment } from '../../../environments/environment';
import { firstValueFrom } from 'rxjs';

const SESSION_KEY = 'ef_session';
const TOKEN_KEY = 'ef_token';

@Injectable({ providedIn: 'root' })
export class AuthService {
  private http = inject(HttpClient);
  private router = inject(Router);
  
  private _currentUser = signal<User | null>(null);

  readonly currentUser  = this._currentUser.asReadonly();
  readonly isLoggedIn   = computed(() => !!this._currentUser());
  readonly isEmployee   = computed(() => this._currentUser()?.role === 'employee');
  readonly isManager    = computed(() => this._currentUser()?.role === 'manager');
  readonly userFullName = computed(() => {
    const u = this._currentUser();
    return u ? `${u.fname} ${u.lname}` : '';
  });
  readonly userInitials = computed(() => {
    const u = this._currentUser();
    return u ? `${u.fname[0]}${u.lname[0]}`.toUpperCase() : '';
  });

  constructor() {
    this.restoreSession();
  }

  // ── Public API ────────────────────────────────────
  async login(email: string, password: string, role: UserRole): Promise<{ success: boolean; error?: string }> {
    try {
      const response: any = await firstValueFrom(
        this.http.post(`${environment.apiUrl}/Auth/login`, {
          email,
          password,
          role
        })
      );

      if (response.success && response.data) {
        const user: User = {
          id: response.data.user.id,
          fname: response.data.user.fname,
          lname: response.data.user.lname,
          email: response.data.user.email,
          password: '',
          role: response.data.user.role as UserRole,
          department: response.data.user.department,
          createdAt: response.data.user.createdAt
        };
        
        this.setSession(user, response.data.token);
        return { success: true };
      }
      
      return { success: false, error: 'Login failed' };
    } catch (error: any) {
      return { 
        success: false, 
        error: error.error?.error || 'Invalid credentials or wrong role selected.' 
      };
    }
  }

  async register(data: Omit<User, 'id' | 'createdAt'>): Promise<{ success: boolean; error?: string }> {
    try {
      const response: any = await firstValueFrom(
        this.http.post(`${environment.apiUrl}/Auth/register`, {
          fname: data.fname,
          lname: data.lname,
          email: data.email,
          password: data.password,
          role: data.role,
          department: data.department
        })
      );

      if (response.success && response.data) {
        return { success: true };
      }
      
      return { success: false, error: 'Registration failed' };
    } catch (error: any) {
      return { 
        success: false, 
        error: error.error?.error || 'Email is already registered.' 
      };
    }
  }

  logout(): void {
    this._currentUser.set(null);
    localStorage.removeItem(SESSION_KEY);
    localStorage.removeItem(TOKEN_KEY);
    this.router.navigate(['/login']);
  }

  // ── Internals ─────────────────────────────────────
  private setSession(user: User, token: string): void {
    this._currentUser.set(user);
    localStorage.setItem(SESSION_KEY, JSON.stringify(user));
    localStorage.setItem(TOKEN_KEY, token);
    const route = user.role === 'employee' ? '/employee/overview' : '/manager/overview';
    this.router.navigate([route]);
  }

  private restoreSession(): void {
    const userJson = localStorage.getItem(SESSION_KEY);
    const token = localStorage.getItem(TOKEN_KEY);
    
    if (userJson && token) {
      try {
        const user = JSON.parse(userJson) as User;
        this._currentUser.set(user);
      } catch {
        this.logout();
      }
    }
  }
}

File: Backend/Backend/Backend/Services/AuthService.cs
using Backend.Models;
using Backend.Models.DTOs;
using Backend.Helpers;
using Backend.Data;
using Microsoft.EntityFrameworkCore;

namespace Backend.Services;

public class AuthService : IAuthService
{
    private readonly ApplicationDbContext _context;
    private readonly JwtHelper _jwtHelper;

    public AuthService(ApplicationDbContext context, JwtHelper jwtHelper)
    {
        _context = context;
        _jwtHelper = jwtHelper;
    }

    public AuthResponse Login(LoginRequest request)
    {
        var role = request.Role.ToLower() == "employee" ? UserRole.Employee : UserRole.Manager;
        var user = _context.Users.FirstOrDefault(u => 
            u.Email == request.Email && 
            u.Role == role);

        if (user == null || !PasswordHelper.VerifyPassword(request.Password, user.Password))
        {
            return new AuthResponse
            {
                Success = false,
                Error = "Invalid credentials or wrong role selected."
            };
        }

        var token = _jwtHelper.GenerateToken(user);

        return new AuthResponse
        {
            Success = true,
            User = MapToDto(user),
            Token = token
        };
    }

    public AuthResponse Register(RegisterRequest request)
    {
        if (_context.Users.Any(u => u.Email == request.Email))
        {
            return new AuthResponse
            {
                Success = false,
                Error = "Email is already registered."
            };
        }

        var user = new User
        {
            Id = "u_" + DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Fname = request.Fname,
            Lname = request.Lname,
            Email = request.Email,
            Password = PasswordHelper.HashPassword(request.Password),
            Role = request.Role.ToLower() == "employee" ? UserRole.Employee : UserRole.Manager,
            Department = request.Department,
            CreatedAt = DateTime.UtcNow
        };

        _context.Users.Add(user);
        _context.SaveChanges();

        var token = _jwtHelper.GenerateToken(user);

        return new AuthResponse
        {
            Success = true,
            User = MapToDto(user),
            Token = token
        };
    }

    public User? GetCurrentUser(string userId)
    {
        return _context.Users.FirstOrDefault(u => u.Id == userId);
    }

    public List<User> GetAllUsers()
    {
        return _context.Users.ToList();
    }

    private UserDto MapToDto(User user)
    {
        return new UserDto
        {
            Id = user.Id,
            Fname = user.Fname,
            Lname = user.Lname,
            Email = user.Email,
            Role = user.Role.ToString().ToLower(),
            Department = user.Department,
            CreatedAt = user.CreatedAt.ToString("o")
        };
    }
}
